File: src/main/kotlin/org/alexn/hook/OperatingSystem.kt
package org.alexn.hook

import arrow.core.Option
import arrow.core.recover
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.runInterruptible
import kotlinx.coroutines.withContext
import org.apache.commons.text.StringEscapeUtils
import java.io.File
import java.nio.charset.StandardCharsets.UTF_8
import java.nio.file.Path

data class CommandResult(
    val exitCode: Int,
    val stdout: String,
    val stderr: String,
) {
    val isSuccessful get() = exitCode == 0
}

/**
 * Executes a program. This needs to be a valid path on the
 * file system.
 *
 * See [executeEscapedShellCommand] for the version that executes
 * `/bin/sh` commands.
 */
suspend fun executeCommand(
    executable: Path,
    args: List<String>? = null,
    dir: File? = null,
): CommandResult =
    // Blocking I/O should use threads designated for I/O
    withContext(Dispatchers.IO) {
        val cmdArgs = listOf(executable.toAbsolutePath().toString()) + (args ?: listOf())
        val proc =
            Runtime.getRuntime().exec(
                cmdArgs.toTypedArray(),
                arrayOf(),
                dir,
            )
        try {
            // Concurrent execution ensures the stream's buffer doesn't
            // block processing when overflowing
            val stdout =
                async {
                    runInterruptible(Dispatchers.IO) {
                        // That `InputStream.read` doesn't listen to thread interruption
                        // signals; but for future development it doesn't hurt
                        String(proc.inputStream.readAllBytes(), UTF_8)
                    }
                }
            val stderr =
                async {
                    runInterruptible(Dispatchers.IO) {
                        String(proc.errorStream.readAllBytes(), UTF_8)
                    }
                }
            CommandResult(
                exitCode = runInterruptible(Dispatchers.IO) { proc.waitFor() },
                stdout = stdout.await(),
                stderr = stderr.await(),
            )
        } finally {
            proc.destroy()
        }
    }

/**
 * Executes shell commands.
 *
 * This version does shell escaping of command arguments.
 * WARN: command arguments need be given explicitly because
 * they need to be properly escaped.

 * @see [executeRawShellCommand]
 */
suspend fun executeEscapedShellCommand(
    command: String,
    args: List<String>? = null,
    dir: File? = null,
): CommandResult =
    executeRawShellCommand(
        command =
            (listOf(command) + (args ?: listOf()))
                .map(StringEscapeUtils::escapeXSI)
                .joinToString(" "),
        dir = dir,
    )

/**
 * Executes shell commands.
 */
suspend fun executeRawShellCommand(
    command: String,
    dir: File? = null,
): CommandResult =
    executeCommand(
        executable = Path.of("/bin/sh"),
        args = listOf("-c", command),
        dir = dir,
    )

val USER_HOME: File? by lazy {
    Option.fromNullable(System.getProperty("user.home"))
        .filter { it.isNotEmpty() }
        .recover { Option.fromNullable(System.getenv("HOME")).bind() }
        .filter { it.isNotEmpty() }
        .map { File(it) }
        .getOrNull()
}

File: src/main/scala/org/alexn/hook/OperatingSystem.scala
package org.alexn.hook

import cats.effect.Outcome
import cats.effect.kernel.Async
import cats.effect.std.Env
import cats.effect.syntax.all.*
import cats.syntax.all.*
import org.apache.commons.text.StringEscapeUtils
import org.typelevel.log4cats
import org.typelevel.log4cats.Logger

import java.io.File
import java.nio.charset.StandardCharsets.UTF_8
import java.nio.file.Path

final case class CommandResult(
    exitCode: Int,
    stdout: String,
    stderr: String
)

class OperatingSystem[F[_]](using Async[F], Env[F], Logger[F]):
    def executeCommand(
        executable: Path,
        args: List[String],
        dir: Option[File]
    ): F[CommandResult] =
        Async[F].blocking:
            val commandArgs = executable.toAbsolutePath.toString +: args
            Runtime.getRuntime.exec(
                commandArgs.toArray,
                Array.empty[String],
                dir.orNull
            )
        .bracketCase: proc =>
            // These aren't "interruptible", what actually interrupts them
            // is proc.destroy(); and due to how they are used, it's better
            // to not declare them as interruptible, as to not mislead:
            val collectStdout = Async[F].blocking:
                new String(proc.getInputStream.readAllBytes(), UTF_8)
            val collectStderr = Async[F].blocking:
                new String(proc.getErrorStream.readAllBytes(), UTF_8)
            for
                // Starts jobs asynchronously
                stdoutFiber <- collectStdout.start
                stderrFiber <- collectStderr.start
                // Waits for process to complete (this is actually interruptible)
                code <- Async[F].interruptible(proc.waitFor())
                // Reads output
                stdout <- stdoutFiber.joinWithNever
                stderr <- stderrFiber.joinWithNever
            yield CommandResult(code, stdout, stderr)
        .apply: (proc, outcome) =>
            for
                _ <- outcome match
                case Outcome.Canceled() => Logger[F].warn("Command was cancelled")
                case Outcome.Errored(e) => Logger[F].error(e)("Command failed")
                case Outcome.Succeeded(_) => Logger[F].debug("Command succeeded")
                _ <- Async[F].blocking(proc.destroy())
            yield ()

    def executeRawShellCommand(
        command: String,
        dir: Option[File]
    ): F[CommandResult] =
        executeCommand(
            executable = Path.of("/bin/sh"),
            args = List("-c", command),
            dir = dir
        )

    def executeEscapedShellCommand(
        command: String,
        args: List[String],
        dir: Option[File]
    ): F[CommandResult] =
        executeRawShellCommand(
            (command +: args).map(StringEscapeUtils.escapeXSI).mkString(" "),
            dir
        )

    def userHome: F[Option[File]] =
        for
            sysProp <- Async[F].delay(Option(System.getProperty("user.home")))
            envVar <- Env[F].get("HOME")
        yield sysProp
            .filter(_.nonEmpty)
            .orElse(envVar)
            .filter(_.nonEmpty)
            .map(File(_))

end OperatingSystem

object OperatingSystem:
    def apply[F[_]](using Async[F], Env[F], Logger[F]): OperatingSystem[F] =
        new OperatingSystem[F]
